### PR TITLE
[Merged by Bors] - feat(CategoryTheory/Bicategory): define adjunctions

### PR DESCRIPTION
**Pull request by yuma-mizuno**
This RP introduces adjunctions and adjoint equivalences in bicategories, and shows that a (non-adjoint) equivalence can be promoted to an adjoint equivalence by upgrading the counit.

The constructor of adjoint equivalences only requires the left triangle identities since I will plan to show that it implies the right triangle identities in a separate PR.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Bicategory/Adjunction.lean [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+Copyright (c) 2023 Yuma Mizuno. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yuma Mizuno
+-/
+import Mathlib.Tactic.CategoryTheory.Coherence
+
+/-!
+# Adjunctions in bicategories
+
+For a 1-morphisms `f : a ⟶ b` and `g : b ⟶ a` in a bicategory, an adjuntion between `f` and `g`
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Bicategory/Adjunction.lean:11]:
```suggestion
For 1-morphisms `f : a ⟶ b` and `g : b ⟶ a` in a bicategory, an adjuntion between `f` and `g`
```

File: Mathlib/CategoryTheory/Bicategory/Adjunction.lean [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+consists of a pair of 2-morphism `η : 𝟙 a ⟶ f ≫ g` and `ε : g ≫ f ⟶ 𝟙 b` satisfying the triangle
+identities. The 2-morphism `η` is called the unit and `ε` is called the counit.
+
+## Main definitions
+
+* `Bicategiry.Adjunction`: adjunctions between two 1-morphisms.
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Bicategory/Adjunction.lean:17]:
```suggestion
* `Bicategory.Adjunction`: adjunctions between two 1-morphisms.
```

File: Mathlib/CategoryTheory/Bicategory/Adjunction.lean [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+## Implementation notes
+
+The computation of 2-morphisms in the proof is done using `calc` blocks. Typically,
+the LHS and the RHS in each step of `calc` are related by simple rewriting up to associators
+and unitors. So the proof for each step should be of the form `rw [...]; coherence`. In practice,
+our proofs look like `rw [...]; simp [bicategoricalComp]; coherence`. The `simp` is not strictly
+necessary, but it speeds up the proof and allow us to avoid increasing the `maxHeartbeats`.
+The speedup is probably due to reducing the length of the expression e.g. by absorbing
+identity maps or applying the pentagon relation. Such a hack may not be necessary if the
+coherence tactic are improved. One possible way would be to perform a such simplification in the
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Bicategory/Adjunction.lean:31]:
```suggestion
coherence tactic is improved. One possible way would be to perform such a simplification in the
```

File: Mathlib/CategoryTheory/Bicategory/Adjunction.lean [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+Copyright (c) 2023 Yuma Mizuno. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yuma Mizuno
+-/
+import Mathlib.Tactic.CategoryTheory.Coherence
+
+/-!
+# Adjunctions in bicategories
+
+For 1-morphisms `f : a ⟶ b` and `g : b ⟶ a` in a bicategory, an adjuntion between `f` and `g`
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Bicategory/Adjunction.lean:11]:
```suggestion
For 1-morphisms `f : a ⟶ b` and `g : b ⟶ a` in a bicategory, an adjunction between `f` and `g`
```

**Comment by joelriou** [comment]
I cannot possibly comment on the tactic code, but otherwise it looks good to me!

File: Mathlib/Tactic/CategoryTheory/Coherence.lean [hunk 5]
```diff
@@ -170,17 +170,19 @@ def monoidalComp {W X Y Z : C} [LiftObj X] [LiftObj Y]
...
 inserting unitors and associators between as necessary. -/
 noncomputable def monoidalIsoComp {W X Y Z : C} [LiftObj X] [LiftObj Y]
     [MonoidalCoherence X Y] (f : W ≅ X) (g : Y ≅ Z) : W ≅ Z :=
   f ≪≫ asIso MonoidalCoherence.hom ≪≫ g
 
-@[inherit_doc monoidalIsoComp]
-infixr:80 " ≪⊗≫ " => monoidalIsoComp -- type as \ot \gg
+@[inherit_doc Mathlib.Tactic.Coherence.monoidalIsoComp]
+scoped[CategoryTheory.MonoidalCategory] infixr:80 " ≪⊗≫ " =>
+  Mathlib.Tactic.Coherence.monoidalIsoComp -- type as \ot \gg
```
Comment (joelriou) [inline Mathlib/Tactic/CategoryTheory/Coherence.lean:185]:
```suggestion
scoped[CategoryTheory.MonoidalCategory] infixr:80 " ≪⊗≫ " =>
  Mathlib.Tactic.Coherence.monoidalIsoComp -- type as \ll \ot \gg
```

**Comment by joelriou** [comment]
bors merge

**Pull request closed without merging**